File: typescript/src/analysis/dispatch.ts
import type { AST, Expression } from "../types.js";
import type { FirstNullable } from "./first-sets.js";
import { CharSet } from "./charset.js";
import { exprFirstSet, exprIsNullable } from "./first-sets.js";

// --- FIRST set conflict detection ---

/**
 * A FIRST set conflict between two branches of an alternation.
 */
export interface FirstSetConflict {
    /** 0-based index of the first conflicting branch. */
    branchA: number;
    /** 0-based index of the second conflicting branch. */
    branchB: number;
    /** The overlapping characters. */
    overlap: CharSet;
}

/**
 * Find FIRST set conflicts in alternation rules.
 *
 * For each rule whose expression is an alternation, computes per-branch FIRST
 * sets and checks for pairwise overlap. Returns a map from rule name to the
 * list of conflicts found.
 */
export function findFirstSetConflicts(
    ast: AST,
    firstNullable: FirstNullable,
): Map<string, FirstSetConflict[]> {
    const conflicts = new Map<string, FirstSetConflict[]>();

    for (const [name, rule] of ast) {
        const expr = rule.expression;
        if (expr.type !== "alternation") continue;

        const branches = expr.value as Expression[];
        if (branches.length < 2) continue;

        // Short-circuit: if rule-level FIRST set has <=1 character, trivially no conflict.
        const ruleFirst = firstNullable.firstSets.get(name);
        if (ruleFirst && ruleFirst.len() <= 1) continue;

        // Compute per-branch FIRST sets.
        const branchFirsts = branches.map((branch) =>
            exprFirstSet(branch, firstNullable.firstSets, firstNullable.nullable, ast),
        );

        // Running union optimization: check against union of prior branches.
        const ruleConflicts: FirstSetConflict[] = [];
        const runningUnion = new CharSet();

        for (let i = 0; i < branchFirsts.length; i++) {
            if (i > 0 && branchFirsts[i].isDisjoint(runningUnion)) {
                // Disjoint with all prior branches -- no conflicts possible.
                runningUnion.union(branchFirsts[i]);
                continue;
            }

            for (let j = 0; j < i; j++) {
                const overlap = branchFirsts[i].intersection(branchFirsts[j]);
                if (!overlap.isEmpty()) {
                    ruleConflicts.push({
                        branchA: j,
                        branchB: i,
                        overlap,
                    });
                }
            }
            runningUnion.union(branchFirsts[i]);
        }

        if (ruleConflicts.length > 0) {
            conflicts.set(name, ruleConflicts);
        }
    }

    return conflicts;
}

// --- Dispatch table ---

export interface DispatchTable {
    table: Int8Array; // 128 entries: charCode -> alternative index, -1 = no match
    isPerfect: boolean; // true if all alternatives are covered disjointly
}

/**
 * Build a dispatch table for an alternation node. Returns null if any
 * alternative is nullable or has an empty/unknown FIRST set.
 */
export function buildDispatchTable(
    alternatives: Expression[],
    firstSets: Map<string, CharSet>,
    nullable: Map<string, boolean>,
): DispatchTable | null {
    const altFirstSets: CharSet[] = [];

    for (const alt of alternatives) {
        // Reject if any alternative is nullable
        if (exprIsNullable(alt, nullable, new Map())) return null;

        const cs = exprFirstSet(alt, firstSets, nullable, new Map());
        if (cs.isEmpty()) return null;
        altFirstSets.push(cs);
    }

    // Check pairwise disjointness
    for (let i = 0; i < altFirstSets.length; i++) {
        for (let j = i + 1; j < altFirstSets.length; j++) {
            if (!altFirstSets[i].isDisjoint(altFirstSets[j])) {
                return null;
            }
        }
    }

    // Build table
    const table = new Int8Array(128).fill(-1);
    for (let i = 0; i < altFirstSets.length; i++) {
        for (let ch = 0; ch < 128; ch++) {
            if (altFirstSets[i].has(ch)) {
                table[ch] = i;
            }
        }
    }

    return { table, isPerfect: true };
}

File: rust/bbnf/src/analysis/dispatch.rs
//! FIRST set conflict detection for alternation rules.

use std::collections::HashMap;

use super::charset::CharSet;
use super::first_sets::{FirstSets, compute_expr_first, unwrap_rule};
use crate::types::{AST, Expression};

// NOTE: DispatchTable and build_dispatch_table were removed — AST-level dispatch
// is dead code. The IR pass `generate_dispatch_tables` handles dispatch.

/// A FIRST set conflict between two branches of an alternation.
#[derive(Debug, Clone)]
pub struct FirstSetConflict {
    pub branch_a: usize,
    pub branch_b: usize,
    pub overlap: CharSet,
}

/// Find FIRST set conflicts in alternation rules.
pub fn find_first_set_conflicts<'a>(
    ast: &'a AST<'a>,
    first_sets: &FirstSets<'a>,
) -> HashMap<String, Vec<FirstSetConflict>> {
    let name_to_key: HashMap<&str, &'a Expression<'a>> = ast
        .keys()
        .filter_map(|lhs| match lhs {
            Expression::Nonterminal(tok) => Some((tok.value.as_ref(), lhs)),
            _ => None,
        })
        .collect();

    let mut conflicts = HashMap::new();

    for (lhs, rhs) in ast {
        let name = match lhs {
            Expression::Nonterminal(tok) => tok.value.to_string(),
            _ => continue,
        };

        let inner = unwrap_rule(rhs);
        let branches = match inner {
            Expression::Alternation(tok) => &tok.value[..],
            _ => continue,
        };

        if branches.len() < 2 {
            continue;
        }

        if let Some(rule_first) = first_sets.first.get(lhs) {
            if rule_first.len() <= 1 {
                continue;
            }
        }

        let branch_firsts_vec: Vec<(CharSet, bool)>;
        let branch_firsts: &[(CharSet, bool)] =
            if let Some(cached) = first_sets.branch_firsts.get(lhs) {
                cached
            } else {
                branch_firsts_vec = branches
                    .iter()
                    .map(|branch| {
                        let mut cs = CharSet::new();
                        let is_nullable = compute_expr_first(
                            branch,
                            &first_sets.first,
                            &first_sets.nullable,
                            &name_to_key,
                            &mut cs,
                        );
                        (cs, is_nullable)
                    })
                    .collect();
                &branch_firsts_vec
            };

        let mut rule_conflicts = Vec::new();
        let mut union_so_far = CharSet::new();

        for i in 0..branch_firsts.len() {
            let (ref branch_i_first, _) = branch_firsts[i];

            if i > 0 && branch_i_first.is_disjoint(&union_so_far) {
                union_so_far.union(branch_i_first);
                continue;
            }

            for (j, (branch_j_first, _)) in branch_firsts.iter().enumerate().skip(i + 1) {
                let overlap = branch_i_first.intersection(branch_j_first);
                if !overlap.is_empty() {
                    rule_conflicts.push(FirstSetConflict {
                        branch_a: i,
                        branch_b: j,
                        overlap,
                    });
                }
            }

            union_so_far.union(branch_i_first);
        }

        if !rule_conflicts.is_empty() {
            conflicts.insert(name, rule_conflicts);
        }
    }

    conflicts
}
